Src/Extern/HYPRE/AMReX_Habec_2D_K.H

```diff
@@ -0,0 +1,673 @@
+#ifndef AMREX_HABEC_2D_H_
+#define AMREX_HABEC_2D_H_
+
+#include <AMReX_LO_BCTYPES.H>
+#ifdef AMREX_USE_HYPRE
+#include <AMReX_Hypre.H>
+#endif
+#include <AMReX_MultiFabUtil.H>
+
+#ifdef AMREX_USE_EB
+#include <AMReX_EBMultiFabUtil.H>
+#include <AMReX_MultiCutFab.H>
+#include <AMReX_EBFabFactory.H>
+#include <AMReX_MLEBABecLap_K.H>
+#endif
+
+namespace amrex {
+
+AMREX_FORCE_INLINE
+void amrex_hpbcoef (Box const& box,
+                    BaseFab<GpuArray<Real,2*AMREX_SPACEDIM + 1>>& mat,
+                    const FArrayBox& b,
+                    Real& sb,
+                    const Real* dx,
+                    int& idim)
+{
+    Array4<GpuArray<Real,2*AMREX_SPACEDIM + 1>> const& mat_arr = mat.array();
+    Array4<Real const> const& b_arr = b.const_array();
+    Real fac = sb / (dx[idim]*dx[idim]);
+
+    if (idim == 0) {
+        AMREX_HOST_DEVICE_PARALLEL_FOR_3D (box, i, j, k,
+        {
+            mat_arr(i,j,k)[0] = mat_arr(i,j,k)[0] + fac * (b_arr(i,j,k) + b_arr(i+1,j,k));
+            mat_arr(i,j,k)[1] = - fac * b_arr(i,j,k);
+            mat_arr(i,j,k)[2] = - fac * b_arr(i+1,j,k);
+        });
+    }else{
+        AMREX_HOST_DEVICE_PARALLEL_FOR_3D (box, i, j, k,
+        {
+            mat_arr(i,j,k)[0] = mat_arr(i,j,k)[0] + fac * (b_arr(i,j,k) + b_arr(i,j+1,k));
+            mat_arr(i,j,k)[3] = - fac * b_arr(i,j,k);
+            mat_arr(i,j,k)[4] = - fac * b_arr(i,j+1,k);
+        });
+    }
+}
+
+AMREX_FORCE_INLINE
+void amrex_hpmat (Box const& box,
+                  BaseFab<GpuArray<Real,2*AMREX_SPACEDIM + 1>>& mat,
+                  const FArrayBox& b,
+                  const Mask& msk,
+                  Real& sb,
+                  const Real* dx,
+                  int& cdir,
+                  const int& bct,
+                  const Real& bcl,
+                  const int& bho)
+{
+    const auto lo = amrex::lbound(box);
+    const auto hi = amrex::ubound(box);
+    Array4<GpuArray<Real,2*AMREX_SPACEDIM + 1>> const& mat_arr = mat.array();
+    Array4<Real const> const& b_arr = b.const_array();
+    Array4<int const> const& msk_arr = msk.const_array();
+
+    Real h = 0.0;
+    if (cdir == 0 || cdir == 2) {
+        h = dx[0];
+    }else{
+        h = dx[1];
+    }
+    Real fac = sb/(h*h);
+
+    Real h2 = 0.0;
+    Real h3 = 0.0;
+    Real bf1 = 0.0;
+    Real bf2 = 0.0;
+    if (bct == AMREX_LO_DIRICHLET) {
+        h2 = 0.5*h;
+        if (bho >= 1) {
+            h3 = 3.0*h2;
+            bf1 = fac*((h3 - bcl)/(bcl + h2) - 1.0);
+            bf2 = fac*(bcl - h2)/(bcl + h3);
+        }else{
+            bf1 = fac*( h/(bcl + h2) - 1.0);
+            bf2 = 0.0;
+        }
+    }else if (bct == AMREX_LO_NEUMANN) {
+        bf1 = -fac;
+        bf2 = 0.0;
+    }else{
+        Abort("hpmat: unsupported boundary type");
+    }
+
+    if (cdir == 0){
+        AMREX_HOST_DEVICE_PARALLEL_FOR_3D (box, i, j, k,
+        {
+            if (msk_arr(lo.x-1,j,k) > 0 && i==lo.x){
```

Comment:
```suggestion
        Box tbx = box;
        tbx.setSmall(0,lo.x).setBig(0,lo.x); // 0: x-direction
        AMREX_HOST_DEVICE_PARALLEL_FOR_3D (tbx, i, j, k,
        {
```
The looping is on a slice only.  Could you fix in other place too including 3D?